lib/ace/mode/markdown_highlight_rules.js

```diff
@@ -110,6 +110,14 @@ var MarkdownHighlightRules = function() {
         }, { // link by reference
             token : ["text", "string", "text", "constant", "text"],
             regex : "(\\[)(" + escaped("]") + ")(\\]\s*\\[)("+ escaped("]") + ")(\\])"
+        }, { // link image 
+            token : ["text", "string", "text", "markup.underline", "string", "text"],
+            regex : "(\\!\\[)(" +                                     // ![
```

Comment:
it would be better to add `!?` to the link url, to not repeat the same regexp twice.